tracker/progress.go

```diff
@@ -169,6 +176,19 @@ func (pr *Progress) SentEntries(entries int, bytes uint64) {
 	}
 }
 
+// CanBumpCommit returns true if sending the given commit index can potentially
+// advance the follower's commit index.
+func (pr *Progress) CanBumpCommit(index uint64) bool {
+	return pr.pendingCommit < min(index, pr.Next-1)
+}
+
+// SentCommit updates the pendingCommit.
+func (pr *Progress) SentCommit(commit uint64) {
+	if commit > pr.pendingCommit {
+		pr.pendingCommit = min(commit, pr.Next-1)
```

Comment:
There is no strict requirement that `pendingCommit` does not regresses. The `pendingCommit` means there is an in-flight message with this commit index. But since there is no guarantee that this message is delivered, we may need to send this commit again.

`pendingCommit` is the equivalent of `Next`, but for commit indices. Both may regress for same reasons (when we have no acks from the follower). I chose to maintain `pendingCommit < Next` invariant. Given this invariant, we have:

`pr.pendingCommit < commit` and `pr.pendingCommit < pr.Next`. This line bumps the `pendingCommit` but no higher than `pr.Next - 1`, to maintain the invariant.

Should probably add some commenting here.